### PR TITLE
[Enhancement][fix] Add a rewrite rule to optimize InPredicate.

### LINKED ISSUE #9738: [Bug] InPredicate child expressions type conversion related exceptions
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

1.0

### What's Wrong?

1.  An exception occurs when the binary predicate and the in predicate of the same column are included in the query condition, and the literal in the in predicate is type-converted.
```
mysql> select * from cost2 where id = 2 or id in ("1");
ERROR 1105 (HY000): errCode = 2, detailMessage = Unexpected exception: org.apache.doris.analysis.CastExpr cannot be cast to org.apache.doris.analysis.LiteralExpr
```

2. Cannot hit tablet using string literal in in predicate.
```
mysql> explain select * from cost2 where id in ("1");
+-----------------------------------------------------------------------------------------------+
| Explain String                                                                                |
+-----------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                               |
|  OUTPUT EXPRS:`default_cluster:testdb`.`cost2`.`id` | `default_cluster:testdb`.`cost2`.`cost` |
|   PARTITION: UNPARTITIONED                                                                    |
|                                                                                               |
|   RESULT SINK                                                                                 |
|                                                                                               |
|   1:EXCHANGE                                                                                  |
|                                                                                               |
| PLAN FRAGMENT 1                                                                               |
|  OUTPUT EXPRS:                                                                                |
|   PARTITION: HASH_PARTITIONED: `default_cluster:testdb`.`cost2`.`id`                          |
|                                                                                               |
|   STREAM DATA SINK                                                                            |
|     EXCHANGE ID: 01                                                                           |
|     UNPARTITIONED                                                                             |
|                                                                                               |
|   0:OlapScanNode                                                                              |
|      TABLE: cost2                                                                             |
|      PREAGGREGATION: OFF. Reason: No AggregateInfo                                            |
|      PREDICATES: `id` IN ('1')                                                                |
|      partitions=1/1                                                                           |
|      rollup: cost2                                                                            |
|      tabletRatio=8/8                                                                          |
|      tabletList=17010,17012,17014,17016,17018,17020,17022,17024                               |
|      cardinality=1                                                                            |
|      avgRowSize=1570.0                                                                        |
|      numNodes=1                                                                               |
+-----------------------------------------------------------------------------------------------+
27 rows in set (0.05 sec)

mysql> explain select * from cost2 where id in (1);
+-----------------------------------------------------------------------------------------------+
| Explain String                                                                                |
+-----------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                               |
|  OUTPUT EXPRS:`default_cluster:testdb`.`cost2`.`id` | `default_cluster:testdb`.`cost2`.`cost` |
|   PARTITION: HASH_PARTITIONED: `default_cluster:testdb`.`cost2`.`id`                          |
|                                                                                               |
|   RESULT SINK                                                                                 |
|                                                                                               |
|   0:OlapScanNode                                                                              |
|      TABLE: cost2                                                                             |
|      PREAGGREGATION: OFF. Reason: No AggregateInfo                                            |
|      PREDICATES: `id` IN (1)                                                                  |
|      partitions=1/1                                                                           |
|      rollup: cost2                                                                            |
|      tabletRatio=1/8                                                                          |
|      tabletList=17024                                                                         |
|      cardinality=1                                                                            |
|      avgRowSize=1570.0                                                                        |
|      numNodes=1                                                                               |
+-----------------------------------------------------------------------------------------------+
17 rows in set (0.05 sec)
```

### What You Expected?

1. 
```
mysql> select * from cost2 where id = 2 or id in ("1");
+------+------+
| id   | cost |
+------+------+
|    1 |  100 |
+------+------+
1 row in set (0.07 sec)
```

2.
```
mysql> explain select * from cost2 where id in ("1");
+-----------------------------------------------------------------------------------------------+
| Explain String                                                                                |
+-----------------------------------------------------------------------------------------------+
| PLAN FRAGMENT 0                                                                               |
|  OUTPUT EXPRS:`default_cluster:testdb`.`cost2`.`id` | `default_cluster:testdb`.`cost2`.`cost` |
|   PARTITION: HASH_PARTITIONED: `default_cluster:testdb`.`cost2`.`id`                          |
|                                                                                               |
|   RESULT SINK                                                                                 |
|                                                                                               |
|   0:OlapScanNode                                                                              |
|      TABLE: cost2                                                                             |
|      PREAGGREGATION: OFF. Reason: No AggregateInfo                                            |
|      PREDICATES: `id` IN (1)                                                                  |
|      partitions=1/1                                                                           |
|      rollup: cost2                                                                            |
|      tabletRatio=1/8                                                                          |
|      tabletList=17024                                                                         |
|      cardinality=1                                                                            |
|      avgRowSize=1570.0                                                                        |
|      numNodes=1                                                                               |
+-----------------------------------------------------------------------------------------------+
17 rows in set (0.04 sec)
```

### How to Reproduce?

1. create table
```
CREATE TABLE `cost2` (
  `id` bigint(20) NULL COMMENT "",
  `cost` bigint(20) SUM NULL COMMENT ""
) ENGINE=OLAP
AGGREGATE KEY(`id`)
COMMENT "OLAP"
DISTRIBUTED BY HASH(`id`) BUCKETS 8
PROPERTIES (
"replication_allocation" = "tag.location.default: 1",
"in_memory" = "false",
"storage_format" = "V2"
)
```

2. insert date
```
insert into cost2 values(1, 100);
```

3. query
```
select * from cost2 where id = 2 or id in ("1");

explain select * from cost2 where id in ("1");
```

### Anything Else?

In the first sql, both column `id` and literal `2` in expression `id = 2` will be converted to column type `bigint`, and both column `id` and literal `"2"` in expression `id in ("1")` will be converted to `varchar`. When the `ColumnRange` is generated, the `id` types in the two expressions are inconsistent, causing an exception.

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by luozenglin**
1. Convert child expressions in InPredicate to column type and discard child expressions in them that cannot be converted exactly.
2. Fix the bug of ColumnRange exception caused by InPredicate child expressions type conversion.
3. Fix the problem that the tablet could not be hit due caused by InPredicate child expressions type conversion.

# Proposed changes

Issue Number: close #9738 

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (No)
2. Has unit tests been added: (Yes)
3. Has document been added or modified: (No Need)
4. Does it need to update dependencies: (No)
5. Are there any changes that cannot be rolled back: (No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/RewriteInPredicateRule.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+public class RewriteInPredicateRule implements ExprRewriteRule {
+    public static ExprRewriteRule INSTANCE = new RewriteInPredicateRule();
+
+    @Override
+    public Expr apply(Expr expr, Analyzer analyzer, ClauseType clauseType) throws AnalysisException {
+        if (!(expr instanceof InPredicate)) {
+            return expr;
+        }
+        InPredicate inPredicate = (InPredicate) expr;
+        if (inPredicate.contains(Subquery.class) || !inPredicate.isLiteralChildren() || inPredicate.isNotIn() ||
```
Comment (jackwener) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/RewriteInPredicateRule.java:57]: Please notice that `checkstyle`, you can see this [doc](https://doris.apache.org/zh-CN/developer-guide/java-format-code.html#import-order)

Comment (luozenglin) [reply fe/fe-core/src/main/java/org/apache/doris/rewrite/RewriteInPredicateRule.java]: done. Thank you.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/RewriteInPredicateRule.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+        List<Expr> invalidChildren = Lists.newArrayList();
+        for (int i = 1; i < newInPredicate.getChildren().size(); ++i) {
+            LiteralExpr childExpr = (LiteralExpr) newInPredicate.getChild(i);
+            if (childExpr.getType().getPrimitiveType().equals(columnType.getPrimitiveType())) {
+                continue;
+            }
+
+            if (childExpr.getType().getPrimitiveType().isCharFamily() || childExpr.getType().isFloatingPointType()) {
+                try {
+                    childExpr = (LiteralExpr) childExpr.castTo(Type.DECIMALV2);
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/RewriteInPredicateRule.java:77]: From the previous logic, the loop will only be entered when the column type is integer family. 
So, should the cast here be converted directly to the column type, not decimal ?

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/RewriteInPredicateRule.java [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+                    newInPredicate.setChild(i, childExpr);
+                    invalidChildren.add(childExpr);
+                    continue;
+                }
+            }
+
+            if (childExpr.getType().isNumericType()) {
+                try {
+                    LiteralExpr newExpr = (LiteralExpr) childExpr.castTo(columnType);
+                    newExpr.checkValueValid();
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/RewriteInPredicateRule.java:88]: The cast to function will only return IntLiteral or LargeIntLiteral.
So the only possibility of data overflow is when Largeint is converted to Int.
Obviously, the checkValueValid function has already been called when Largeint cast to int, so there is no need to check again here.

File: fe/fe-core/src/main/java/org/apache/doris/rewrite/RewriteInPredicateRule.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+            }
+        }
+        if (invalidChildren.size() == newInPredicate.getChildren().size() - 1) {
+            return new BoolLiteral(false);
+        }
+        if (newInPredicate.getChild(0).getType().getPrimitiveType() != columnType.getPrimitiveType()) {
+            newInPredicate.castChild(columnType, 0);
+            isCast = true;
+        }
+        newInPredicate.getChildren().removeAll(invalidChildren);
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/rewrite/RewriteInPredicateRule.java:110]: Why didn't the illegal literal not be added to new expr directly in the front?

File: fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java [hunk 5]
```diff
@@ -302,6 +302,7 @@ public boolean isAnalyzed() {
         return isAnalyzed;
     }
 
+    // todo: Why define checkValueValid() in Expr that are only implemented in IntLiteral？
```
Comment (EmmyMiao87) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/Expr.java:305]: Remove this comment ~

**Review by EmmyMiao87 — APPROVED** [review]
LGTM

**Pull request merged**